### PR TITLE
[REFAC] simplify derive package

### PR DESCRIPTION
**Pull request by NDStrahilevitz**
## Initial Checklist

- [x] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

```
Author: Nadav Strahilevitz <nadav.strahilevitz@aquasec.com>
Date:   Wed Aug 24 13:48:19 2022 +0000

    events/derive: simplify files
    
    Refactor includes multiple simplifcations to the event derivation code:
    1. Merge events/derive.go and derive_test.go to events/derive/derive.go.
    2. Simplify and rewrite names and comments across the package.
    3. Move errors to an errors.go file
```

Fixes: #2018 

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [x] Code refactor (code improvement and/or code removal)
- [ ] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

No changes were done but tests were moved from:
`pkg/events/derive_test.go` -> `pkg/events/derive/derive_test.go`

For confirmation, I did a sanity check with container_create (docker run --rm alpine) and hooked_syscalls (insmod diamorphine).

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [ ] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: pkg/events/derive/derive.go [hunk 1]
```diff
@@ -1,55 +1,81 @@
...
 import (
-	"fmt"
-
 	"github.com/aquasecurity/tracee/pkg/events"
 	"github.com/aquasecurity/tracee/types/trace"
 )
 
-// Use as static variable for testability reasons
-var getEventDefFunc = events.Definitions.Get
+// deriveFn is a function prototype for a function that receives an event as
```
Comment (rafaeldtinoco) [inline pkg/events/derive/derive.go:8]: nit: `deriveFunction` instead of `deriveFn`

File: pkg/events/derive/derive.go [hunk 2]
```diff
@@ -1,55 +1,81 @@
...
+// argument and may produce a new event if relevant.
+// It returns a derived or empty event, depending on successful derivation,
+// and an error if one occurred.
+type deriveFunction func(trace.Event) ([]trace.Event, []error)
+
+// derive.Table defines a table between events and events they can be derived into corresponding to a deriveFunction
```
Comment (rafaeldtinoco) [inline pkg/events/derive/derive.go:14]: nit: remove `derive.`

File: pkg/events/derive/derive.go [hunk 3]
```diff
@@ -1,55 +1,81 @@
...
+// An Enabled flag is used in order to skip derivation of unwanted events.
```
Comment (rafaeldtinoco) [inline pkg/events/derive/derive.go:15]: nit: remote this line or the // after bool (they're redundant).

**Review by rafaeldtinoco — APPROVED** [review]
LGTM (please fix the 3 comments only before merging).

**Review by yanivagman — CHANGES_REQUESTED** [review]
I see that pkg/ebpf/events_derived.go still exists although in #2018 it says:
`2. Move the pipeline stage from events_derived.go into events_pipeline.go (remove events_derived after)`
Same for:
`3. Move initDerivationTable function to events/derive package (need to pass dependencies in arguments since we remove it from being a method) `

**Comment by NDStrahilevitz** [comment]
> I see that pkg/ebpf/events_derived.go still exists although in #2018 it says: `2. Move the pipeline stage from events_derived.go into events_pipeline.go (remove events_derived after)` Same for: `3. Move initDerivationTable function to events/derive package (need to pass dependencies in arguments since we remove it from being a method) `

I'm personally not so sure about these requirements atm, I think with these changes it's much clearer where the mechanism's interfaces and logic lies.
Then tracee-ebpf simply consumes this logic, creating it's own derivation table.
Do you find that even with this change the code is still too spread?

**Comment by rafaeldtinoco** [comment]
@yanivagman I had +1'ed this, and you asked more info. Should we merge or change something before ?

**Comment by yanivagman** [comment]
> > I see that pkg/ebpf/events_derived.go still exists although in #2018 it says: `2. Move the pipeline stage from events_derived.go into events_pipeline.go (remove events_derived after)` Same for: `3. Move initDerivationTable function to events/derive package (need to pass dependencies in arguments since we remove it from being a method) `
> 
> I'm personally not so sure about these requirements atm, I think with these changes it's much clearer where the mechanism's interfaces and logic lies. Then tracee-ebpf simply consumes this logic, creating it's own derivation table. Do you find that even with this change the code is still too spread?

Yes, I think that at least (2) should be part of this PR, since it is the only pipeline stage that is currently not part of events_pipeline.go

**Comment by NDStrahilevitz** [comment]
> Yes, I think that at least (2) should be part of this PR, since it is the only pipeline stage that is currently not part of events_pipeline.go

Ok, i'll add that change.

**Comment by NDStrahilevitz** [comment]
@yanivagman Added change from (2), WDYT? I think keeping the derivation table in a separate file might be better but I'm fine with leaving it like this.

File: pkg/events/derive/errors.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+	"github.com/aquasecurity/tracee/pkg/events"
+)
+
+func deriveError(id events.ID, err error) error {
+	return fmt.Errorf("failed to derive event %d: %v", id, err)
+}
+
+func unexpectedArgCountError(name string, expected int, actual int) error {
+	return fmt.Errorf("error deriving event \"%s\": expected %d arguments but given %d", name, expected, actual)
```
Comment (josedonizetti) [inline pkg/events/derive/errors.go:14]: nit
```suggestion
	return fmt.Errorf("error deriving event %q: expected %d arguments but given %d", name, expected, actual)
```

File: pkg/events/derive/derive.go [hunk 5]
```diff
@@ -1,55 +1,81 @@
...
 	"github.com/aquasecurity/tracee/types/trace"
 )
 
-// Use as static variable for testability reasons
-var getEventDefFunc = events.Definitions.Get
+// deriveFunction is a function prototype for a function that receives an event as
+// argument and may produce a new event if relevant.
+// It returns a derived or empty event, depending on successful derivation,
+// and an error if one occurred.
+type deriveFunction func(trace.Event) ([]trace.Event, []error)
```
Comment (josedonizetti) [inline pkg/events/derive/derive.go:12]: @NDStrahilevitz not sure if pertinent to this PR, but all derive functions return a single error `[]tracee.Event, error`, but we covert it to an slice inside `singleEventDeriveFunc` without needing it. Only `DeriveEvent` needs to return multiple errors. WDYT?

**Review by josedonizetti — COMMENTED** [review]
Nice refactoring. Added a nit and a question.

File: pkg/events/derive/derive.go (hunk 5 above)
Comment (NDStrahilevitz) [reply pkg/events/derive/derive.go:12]: Take a look at #2053, not sure when it will be merged but there is an example for this requirement there.

**Review by josedonizetti — APPROVED** [review]

**Review by yanivagman — APPROVED** [review]
LGTM, thanks for taking care of these changes

**Pull request merged**